subprojects/antlr/src/main/java/org/gradle/api/plugins/antlr/AntlrPlugin.java

```diff
@@ -68,16 +81,20 @@ public void apply(final Project project) {
         // Wire the antlr configuration into all antlr tasks
         project.getTasks().withType(AntlrTask.class).configureEach(antlrTask -> antlrTask.getConventionMapping().map("antlrClasspath", () -> project.getConfigurations().getByName(ANTLR_CONFIGURATION_NAME)));
 
-        project.getExtensions().getByType(JavaPluginExtension.class).getSourceSets().all(
+        JavaPluginExtension javaPluginExtension = project.getExtensions().getByType(JavaPluginExtension.class);
+        javaPluginExtension.getSourceSets().all(
                 new Action<SourceSet>() {
                     @Override
                     public void execute(final SourceSet sourceSet) {
                         // for each source set we will:
                         // 1) Add a new 'antlr' virtual directory mapping
+
+
                         final AntlrSourceVirtualDirectoryImpl antlrDirectoryDelegate
-                                = new AntlrSourceVirtualDirectoryImpl(((DefaultSourceSet) sourceSet).getDisplayName(), objectFactory);
+                                = instantiator.newInstance(AntlrSourceVirtualDirectoryImpl.class, ((DefaultSourceSet) sourceSet).getDisplayName(), objectFactory, patternSetFactory, fileCollectionFactory, directoryFileTreeFactory);
                         new DslObject(sourceSet).getConvention().getPlugins().put(
                                 AntlrSourceVirtualDirectory.NAME, antlrDirectoryDelegate);
+                        sourceSet.getExtensions().add(AntlrSourceVirtualDirectory.NAME, antlrDirectoryDelegate);
```

Comment:
Instead of using `objectFactory`, can't we use `sourceSet.getExtensions().create()`?

Reply:
Same as above.